vendors/espressif/boards/esp32/ports/ble/nimble/iot_ble_hal_gatt_server.c

```diff
@@ -509,63 +403,109 @@ static int prvGATTCharAccessCb( uint16_t conn_handle,
     return rc;
 }
 
+bool setNewHandle( const ble_uuid_t * uuid,
+                   uint16_t handle )
+{
+    bool foundUuid = false;
+    ble_uuid128_t uuid128;
+    uint16_t index = 0;
+    uint16_t serviceIndex = 0;
+    BTService_t * pxService;
+
+    do
+    {
+        pxService = afrServices[ serviceIndex ];
+
+        for( index = 0; index < pxService->xNumberOfAttributes; index++ )
+        {
+            switch( pxService->pxBLEAttributes[ index ].xAttributeType )
+            {
+                case eBTDbPrimaryService:
+                case eBTDbSecondaryService:
+                    ( void * ) prvCopytoESPUUID( &pxService->pxBLEAttributes[ index ].xServiceUUID, ( ble_uuid_t * ) &uuid128 );
+                    break;
+
+                case eBTDbCharacteristic:
+                    ( void * ) prvCopytoESPUUID( &pxService->pxBLEAttributes[ index ].xCharacteristic.xUuid, ( ble_uuid_t * ) &uuid128 );
+                    break;
+
+                case eBTDbDescriptor:
+                    ( void * ) prvCopytoESPUUID( &pxService->pxBLEAttributes[ index ].xCharacteristicDescr.xUuid, ( ble_uuid_t * ) &uuid128 );
+                    break;
+
+                default:
+                    break;
+            }
+
+            /* Complete handle for CCCD. Since no callback are generated for CCCD (because they are added automatically.
+             * The handle of CCCDs is still 0. */
+            if( index + 1 < pxService->xNumberOfAttributes )
+            {
+                if( pxService->pxBLEAttributes[ index + 1 ].xAttributeType == eBTDbDescriptor )
+                {
+                    /* If the attribute is a CCCD then give it the handle of previous attribute + 1.*/
+                    if( ( pxService->pxBLEAttributes[ index + 1 ].xCharacteristicDescr.xUuid.ucType == eBTuuidType16 ) &&
+                        ( pxService->pxBLEAttributes[ index + 1 ].xCharacteristicDescr.xUuid.uu.uu16 == BLE_GATT_DSC_CLT_CFG_UUID16 ) )
+                    {
+                        pxService->pusHandlesBuffer[ index + 1 ] = pxService->pusHandlesBuffer[ index ] + 1;
+                    }
+                }
+            }
+
+            if( ble_uuid_cmp( uuid, ( ble_uuid_t * ) &uuid128 ) == 0 )
+            {
+                if( ( ( pxService->pxBLEAttributes[ index ].xAttributeType == eBTDbPrimaryService ) ||
+                      ( pxService->pxBLEAttributes[ index ].xAttributeType == eBTDbSecondaryService ) ) &&
+                    ( serviceIndex == 0 ) )
+                {
+                    gattOffset = handle;
+                }
+
+                pxService->pusHandlesBuffer[ index ] = handle - gattOffset;
+                break;
+            }
+        }
+
+        serviceIndex++;
+    } while( espServices[ serviceIndex ].type != BLE_GATT_SVC_TYPE_END );
```

Comment:
Bound by MAX_SERVICES  as well ?

Reply:
No, that is why Nimble request the MAX_SERVICE+1.
The last service always has BLE_GATT_SVC_TYPE_END. I had that same comment when I read the code